contrib/isa_l/compression/source/compressor/igzip_compressor_impl.cc

```diff
@@ -0,0 +1,139 @@
+#include "contrib/isa_l/compression/source/compressor/igzip_compressor_impl.h"
+
+#include <memory>
+
+#include "envoy/common/exception.h"
+
+#include "source/common/common/assert.h"
+
+#include "absl/container/fixed_array.h"
+#include "isa-l.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Compression {
+namespace Igzip {
+namespace Compressor {
+
+IgzipCompressorImpl::IgzipCompressorImpl() : IgzipCompressorImpl(4096) {}
+
+IgzipCompressorImpl::IgzipCompressorImpl(uint64_t chunk_size)
+    : Common::Base(chunk_size, [](isal_zstream* z) {
+        free(z->level_buf);
+        delete z;
+      }) {}
+
+void IgzipCompressorImpl::init(CompressionLevel comp_level, int64_t window_bits,
+                               int64_t memory_level) {
+  ASSERT(initialized_ == false);
+  isal_deflate_init(zstream_ptr_.get());
+  zstream_ptr_->avail_out = chunk_size_;
+  zstream_ptr_->next_out = chunk_char_ptr_.get();
+  zstream_ptr_->gzip_flag = IGZIP_GZIP;
+  zstream_ptr_->hist_bits = window_bits;
+
+  if (comp_level == CompressionLevel::Level1) {
+    zstream_ptr_->level = 1;
+  } else if (comp_level == CompressionLevel::Level2) {
+    zstream_ptr_->level = 2;
+  } else {
+    zstream_ptr_->level = 3;
+  }
+
+  uint64_t buf_size = 0;
+  uint64_t joint_index = (zstream_ptr_->level - 1) * 5 + memory_level - 1;
+  switch (joint_index) {
```

Comment:
Is there a way to count final level directly? There are too many cases.

Reply:
I think an array with joint_index as index can avoid the 'switch case'

Reply:
There are five memory levels for each compression level, and the sizes are defined using macros. Mayebe we can skip these macros to calculate relevant size for each memory level.

Reply:
> I think an array with joint_index as index can avoid the 'switch case'

Yeah, I also think it's good to use an array.Thanks!